File: app/src/main/java/com/moggot/findmycarlocation/billing/BillingManager.java
package com.moggot.findmycarlocation.billing;

import android.app.Activity;
import android.support.annotation.Nullable;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.ConsumeResponseListener;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchaseHistoryResponseListener;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.android.billingclient.api.SkuDetailsResponseListener;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class BillingManager {

    public static final int BILLING_MANAGER_NOT_INITIALIZED = -1;
    private final Activity mActivity;
    private BillingClient mBillingClient;
    @Nullable
    private BillingReadyListener mBillingReadyListener;
    private int mBillingClientResponseCode = BILLING_MANAGER_NOT_INITIALIZED;
    private boolean showAds = true;

    @Inject
    public BillingManager(Activity activity) {
        mActivity = activity;
        mBillingClient = BillingClient.newBuilder(activity)
                .setListener(new PurchasesUpdatedListener() {
                    @Override
                    public void onPurchasesUpdated(int responseCode, @Nullable List<Purchase> purchases) {
                        if (responseCode == BillingClient.BillingResponse.OK &&
                                purchases != null) {
                            for (Purchase purchase : purchases) {
                                mBillingClient.consumeAsync(purchase.getPurchaseToken(), new AdsPolicyUpdateListener());
                            }
                        }
                    }
                }).build();
    }

    public int getBillingClientResponseCode() {
        return mBillingClientResponseCode;
    }

    public void startConnection() {
        mBillingClient.startConnection(new BillingClientStateListener() {
            @Override
            public void onBillingSetupFinished(int responseCode) {
                if (responseCode == BillingClient.BillingResponse.OK) {
                    mBillingClient.queryPurchaseHistoryAsync(BillingClient.SkuType.SUBS, new PurchaseHistoryResponseListener() {
                        @Override
                        public void onPurchaseHistoryResponse(int responseCode, List<Purchase> purchases) {
                            if (responseCode == BillingClient.BillingResponse.OK &&
                                    purchases != null) {
                                for (Purchase purchase : purchases) {
                                    mBillingClient.consumeAsync(purchase.getPurchaseToken(), new AdsPolicyUpdateListener());
                                }
                            }
                            if (mBillingReadyListener != null) {
                                mBillingReadyListener.billingReady();
                            }
                        }
                    });
                }
                mBillingClientResponseCode = responseCode;
            }

            @Override
            public void onBillingServiceDisconnected() {
                mBillingClient.startConnection(this);

            }
        });
    }

    public boolean isPremium() {
        return !showAds;
    }

    public void setAdsShowListener(BillingReadyListener billingReadyListener) {
        mBillingReadyListener = billingReadyListener;
    }

    public void requestSubscription() {
        if (!showAds) {
            return;
        }
        List<String> skuList = new ArrayList<String>() {{
            add("ads_disable_subscription");
        }};

        SkuDetailsParams params = SkuDetailsParams.newBuilder()
                .setSkusList(skuList)
                .setType(BillingClient.SkuType.SUBS)
                .build();
        mBillingClient.querySkuDetailsAsync(params, new SkuDetailsResponseListener() {
            @Override
            public void onSkuDetailsResponse(int responseCode, List<SkuDetails> skuDetailsList) {
                if (responseCode == BillingClient.BillingResponse.OK &&
                        skuDetailsList != null) {
                    for (SkuDetails skuDetails : skuDetailsList) {
                        BillingFlowParams.Builder builder = BillingFlowParams.newBuilder()
                                .setSkuDetails(skuDetails);
                        if (mBillingClient.isFeatureSupported(BillingClient.FeatureType.SUBSCRIPTIONS)
                                == BillingClient.BillingResponse.OK) {
                            mBillingClient.launchBillingFlow(mActivity, builder.build());
                        }
                    }
                } else {
                    startConnection();
                }
            }
        });
    }

    public void destroy() {
        if (mBillingClient != null && mBillingClient.isReady()) {
            mBillingClient.endConnection();
        }
    }

    private class AdsPolicyUpdateListener implements ConsumeResponseListener {

        @Override
        public void onConsumeResponse(int responseCode, String purchaseToken) {
            if (responseCode == BillingClient.BillingResponse.OK) {
                return;
            }
            showAds = false;
        }
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/billing/BillingManager.kt
package com.moggot.findmycarlocation.billing

import android.app.Activity
import android.util.Log
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.ProductDetailsResponseListener
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.moggot.findmycarlocation.App.Companion.TAG
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BillingManager @Inject constructor(private val mActivity: Activity) {

    private val _purchases = MutableStateFlow<List<Purchase>>(listOf())
    val isPremium: Flow<Boolean> = _purchases.asStateFlow().map { it.isNotEmpty() }

     private val mBillingClient: BillingClient  = BillingClient.newBuilder(mActivity)
         .enablePendingPurchases()
         .setListener { result, purchases: MutableList<Purchase>? ->
             if (result.responseCode == BillingClient.BillingResponseCode.OK && purchases != null) {
                 _purchases.value = purchases
             }
         }.build()

     var billingClientResponseCode = BILLING_MANAGER_NOT_INITIALIZED
         private set
     fun startConnection() {
         mBillingClient.startConnection(object : BillingClientStateListener {
             override fun onBillingServiceDisconnected() {
                 mBillingClient.startConnection(this)
             }

             override fun onBillingSetupFinished(result: BillingResult) {
                 billingClientResponseCode = result.responseCode
                 if (billingClientResponseCode == BillingClient.BillingResponseCode.OK) {
                     if (!mBillingClient.isReady) {
                         Timber.tag(TAG).e("queryPurchases: BillingClient is not ready")
                     }
                     mBillingClient.queryPurchasesAsync(QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.SUBS).build()) { billingResult, purchaseList ->
                         if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                             if (purchaseList.isNotEmpty()) {
                                 _purchases.value = purchaseList
                             } else {
                                 _purchases.value = emptyList()
                             }
                         } else {
                             Timber.tag(TAG).e(billingResult.debugMessage)
                         }
                     }
                 }
             }
         })
     }

     fun requestSubscription() {
         val params = QueryProductDetailsParams.newBuilder()
         val productList = mutableListOf<QueryProductDetailsParams.Product>()
         val listOfProducts = listOf(SKU_NAME)
         for (product in listOfProducts) {
             productList.add(
                 QueryProductDetailsParams.Product.newBuilder()
                     .setProductId(product)
                     .setProductType(BillingClient.ProductType.SUBS)
                     .build()
             )

             params.setProductList(productList).let { productDetailsParams ->
                 Timber.tag(TAG).i("queryProductDetailsAsync")
                 mBillingClient.queryProductDetailsAsync(productDetailsParams.build(), object : ProductDetailsResponseListener {
                     override fun onProductDetailsResponse(
                         result: BillingResult,
                         skuDetailsList: MutableList<ProductDetails>,
                     ) {
                         val responseCode = result.responseCode
                         if (responseCode == BillingClient.BillingResponseCode.OK) {
                             for (skuDetails: ProductDetails in skuDetailsList) {
                                 val offers: List<ProductDetails.SubscriptionOfferDetails>? = skuDetails.subscriptionOfferDetails
                                 Log.d(TAG, "skuDetails: $skuDetails")
                                 Log.d(TAG, "offers: $offers")
                                 val offerToken = offers?.firstOrNull()
                                 Log.d(TAG, "offerToken: $offerToken")
                                 offerToken?.let { token: ProductDetails.SubscriptionOfferDetails ->
                                     val builder = BillingFlowParams.newBuilder().setProductDetailsParamsList(
                                         listOf(
                                             BillingFlowParams.ProductDetailsParams.newBuilder()
                                                 .setProductDetails(skuDetails)
                                                 .setOfferToken(token.offerToken)
                                                 .build()
                                         )
                                     )
                                     if (mBillingClient.isFeatureSupported(BillingClient.FeatureType.SUBSCRIPTIONS).responseCode
                                         == BillingClient.BillingResponseCode.OK
                                     ) {
                                         mBillingClient.launchBillingFlow(mActivity, builder.build())
                                     }
                                 }
                             }
                         } else {
                             startConnection()
                         }
                         Log.d(TAG, "onProductDetailsResponse: $skuDetailsList")
                     }
                 })
             }
         }
     }

     fun destroy() {
         if (mBillingClient.isReady) {
             mBillingClient.endConnection()
         }
     }

     companion object {
         const val BILLING_MANAGER_NOT_INITIALIZED = -1
         private const val SKU_NAME = "ads_disable_subscription"
     }
}
